### PR TITLE
Fix issue #6003

### PR DESCRIPTION
**Pull request by muahah**
#6003 
In`streamplot`, skip `start_points` that are on an existing streamline instead of raising `InvalidIndexError`.

### CODE REVIEW

**Comment by WeatherGod** [comment]
Should add a unit test to exercise this case.

**Comment by muahah** [comment]
Run to another problem while adding a test case : 
`streamplot()` accept 1d or 2d arrays for `x` and `y` arguments (transformation from 2d to 1d is done by `Grid()`).
But trying to use 2d arrays while specifying `start_points` lead to a `ValueError`.
At a first though, transforming `x` and `y` to 1d arrays at  `streamplot.py:140` should fix the problem.
I will try to do so soon.
Run into another problem while adding a test case : 
`streamplot()` accept 1d or 2d arrays for `x` and `y` arguments (transformation from 2d to 1d done by `Grid()`).
But trying to use 2d arrays while specifying `start_points` lead to a `ValueError`.
At first guess, transforming `x` and `y` to 1d arrays at  `streamplot.py:140` should fix the problem.
I will try to do so soon.

**Comment by tacaswell** [comment]
attn @tonysyu

**Comment by muahah** [comment]
I fixed issue #6003 and added a test case for `streamplot()` with non-null `start_points` argument.
The test case raised an error because of the use of 2-dimensionnal arrays for `x` and `y` (`Grid` class, that normally deal with 2-dimensionnal arrays is not called in that case), so I made some changes for that too.

File: lib/matplotlib/streamplot.py [hunk 1]
```diff
@@ -137,8 +137,14 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
...
-        sp2[:, 0] += np.abs(x[0])
-        sp2[:, 1] += np.abs(y[0])
+        if x.ndim == 1 and y.ndim == 1:
+            sp2[:, 0] += np.abs(x[0])
+            sp2[:, 1] += np.abs(y[0])
+        elif x.ndim == 2 and y.ndim == 2:
+            sp2[:, 0] += np.abs(x[0][0])
+            sp2[:, 1] += np.abs(y[0][0])
+        else:
+            raise ValueError("'x' and 'y' should have the same dimensions")
```
Comment (QuLogic) [inline lib/matplotlib/streamplot.py]: This message is not strictly correct if, e.g., `x.ndim == y.ndim == 3`.

**Comment by muahah** [comment]
By looking more closely, those checks are redundant because `Grid` already checks `x` and `y` (ndim and values).
`Grid` also extract `x` and `y` origins that can be re-used to shift the seed points, instead of getting them from data.

**Comment by QuLogic** [comment]
So is this complete? None of the streamlines in the test image appear to originate at the plotted points.

Also, the file names do not need "test_image" in them; it's clear it's a test image from the directory.

**Comment by muahah** [comment]
I pointed out the fact that streamlines do not exactly originate from starting points in #6002.
Don't know yet if particular seeding positions on the grid are needed by the streamline algorithm or if there is just a problem while passing between data coordinates and grid coordinates.
I will keep looking on that.
I don't know if it belongs to another pull request or not.

Regarding the image names, I can change them.

**Comment by tacaswell** [comment]
Please do not merge the master branch into your branch.  Instead rebase your branch on top of current master and then force-push to your gh repo.

**Comment by muahah** [comment]
I realize that this pull request is a bit messy (I'm slowly getting used to git and github).

Does it sound good to you if we close it so I can open a new (clean) pull request for issues #6002 and #6003 when it will be ready ?

**Comment by tacaswell** [comment]
The easiest thing to do is to fix this is to rebase + force push.  I just did the rebase locally and there is fortunately no conflicts so (I am going to assume that your upsrteam remote is called 'matplotlib' and the remote pointing at your gh account is called origin


```shell
# update your computer to know what gh knows
git remote update  
# make sure you are one the branch this PR looks at
git checkout master
# rebase onto current master
git rebase matplotlib/master
# check gitk to make sure tis went well

# fail to push to your gh for pedagogical reasons 
git push origin master:master
# try again with --force, only use this when you are sure
git push --force origin master:master
```

**Comment by muahah** [comment]
Thank you, this is better now.
I think this is it, I had to change the way the function pass from data coordinates to internal grid coordinates (the bug was here).
After that some tests failed with low RMS (0.001 and 0.072). 
Theoretically, the change made should not modify the result from `streamplot(start_points=None)` but it indeed change the way data are passed from and to grid coordinates. So I assume small differences in tests are due to numerical errors...

**Comment by tacaswell** [comment]
It looks like you added some new test images and then modified them in later commits.  Can you squash these down into a single commit so that the repository does not carry around the (wrong) intermediary results?

**Comment by muahah** [comment]
I rebased the whole thing to have clean an explicit commits (and no intermediary results).

**Comment by tacaswell** [comment]
Thanks!

The lower left point in the start point test does fall exactly on the stream line.  Is that expected?

**Comment by muahah** [comment]
The streamline do not originate from this point but from the point in the very corner (-3; -3).
the point at (-2.33, -2.33) is ignored because too close to an existing streamline (c.f. e9ddd00).
It is possible to get it back by increasing the `density` parameter.

**Comment by tacaswell** [comment]
Ah, I missed that point was there.

**Pull request merged**